### DISCUSSION TITLE
Building with Deno

### DISCUSSION THREAD

**Discussion by AngelaDMerkel** [Homebrew/discussions#5253] [Writing Formulae/Casks] [answered]
### Output of `brew config`

```shell
HOMEBREW_VERSION: 4.2.16
ORIGIN: https://github.com/Homebrew/brew
HEAD: bd1155be8f50998429a795c15a69c8fe75250510
Last commit: 3 hours ago
Core tap HEAD: 133e5ded926749869284ca7fdeb8d18e76f8b03f
Core tap last commit: 49 minutes ago
Core tap JSON: 01 Apr 18:04 UTC
Core cask tap origin: https://github.com/AngelaDMerkel/homebrew-cask.git
Core cask tap HEAD: 5562a8aaa79fda2e53169ce23d30cc2e1991290f
Core cask tap last commit: 5 days ago
Core cask tap branch: single-file
Core cask tap JSON: 01 Apr 18:04 UTC
HOMEBREW_PREFIX: /opt/homebrew
HOMEBREW_CASK_OPTS: []
HOMEBREW_MAKE_JOBS: 12
HOMEBREW_SORBET_RUNTIME: set
Homebrew Ruby: 3.1.4 => /opt/homebrew/Library/Homebrew/vendor/portable-ruby/3.1.4/bin/ruby
CPU: dodeca-core 64-bit arm_blizzard_avalanche
Clang: 15.0.0 build 1500
Git: 2.39.3 => /Library/Developer/CommandLineTools/usr/bin/git
Curl: 8.4.0 => /usr/bin/curl
macOS: 14.2.1-arm64
CLT: 15.1.0.0.1.1700200546
Xcode: N/A
Rosetta 2: false
```

### Output of `brew doctor`

```shell
Warning: Some installed kegs have no formulae!
This means they were either deleted or installed manually.
You should find replacements for the following formulae:
  python-docopt

Warning: Some installed formulae are deprecated or disabled.
You should find replacements for the following formulae:
  python-charset-normalizer
  python-idna
  python-requests
  python-urllib3
  youtube-dl

Warning: You have unlinked kegs in your Cellar.
Leaving kegs unlinked can lead to build-trouble and cause formulae that depend on
those kegs to fail to run properly once built. Run `brew link` on these:
  glib

Warning: You have uncommitted modifications to Homebrew/homebrew-core.
If this is a surprise to you, then you should stash these modifications.
Stashing returns Homebrew to a pristine state but can be undone
should you later need to do so for some reason.
  cd /opt/homebrew/Library/Taps/homebrew/homebrew-core && git stash -u && git clean -d -f

Uncommitted files:
  ?? Formula/2.0.14.rb
  ?? Formula/s/single-file-cli.rb

You have uncommitted modifications to Homebrew/homebrew-cask.
If this is a surprise to you, then you should stash these modifications.
Stashing returns Homebrew to a pristine state but can be undone
should you later need to do so for some reason.
  cd /opt/homebrew/Library/Taps/homebrew/homebrew-cask && git stash -u && git clean -d -f

Uncommitted files:
   M Casks/s/single-file.rb
```

### Description of issue

I am attempting to build [single-file-cli](https://github.com/gildas-lormeau/single-file-cli/releases/tag/v2.0.14) but having some issue with how to build and move the new binary to the correct location on the system. From what I can see I have the dependencies correct.

```
class SingleFileCli < Formula
  desc "CLI tool for saving a faithful copy of a complete web page in a single HTML file"
  homepage "https://github.com/gildas-lormeau/single-file-cli/blob/master/README.MD"
  url "https://github.com/gildas-lormeau/single-file-cli/archive/refs/tags/v2.0.14.tar.gz"
  sha256 "bb12b99dbef208c62cd7733cbbf0930f4d088fc2bcad8fc459142a375c5a9eaa"
  license "AGPL-3.0-or-later"

  depends_on "deno"
  depends_on "jq"

  def install
    system "./configure", "--disable-silent-rules"
  end

  def post_install
    Dir.chdir("./single-file-cli") do
      system "./compile.sh"
      system "find . -type f ! -name 'single-file' -delete"
    end
  end

  test do
    system "false"
  end
end
```

**Comment by SMillerDev** [comment]
Why do all of that in a post_install block?

**Comment by alebcay** [comment, marked as answer]
In Homebrew, all of the installation tasks (including moving the files into the right place) should be completed in the `install` block (this differs from some other package management systems like AUR or ports where there are distinct compile and package steps).

You should move your post-install steps into the install block. For installing the final build products into the appropriate location, you can use methods such as `bin.install "foo/bar/baz"` to install the binary in `<build temp dir>/foo/bar/baz"` as `<install prefix>/bin/baz`.

You can see more about the `install` method (and other general knowledge about writing formulae) in the [Formula Cookbook](https://docs.brew.sh/Formula-Cookbook).

**Reply by AngelaDMerkel** [reply]
Once I have downloaded the tarball, what's preventing from building there and then moving the executable to the correct location as in this following? 

```
  def install
      system "chmod +x compile.sh"
      system "./compile.sh"
      system "find . -type f ! -name 'single-file' -delete"
      bin.install "single-file" => "single-file-cli"
  end
```

Forgive any ignorance, but this seems successful in building the executable and then cleaning up any leftovers, but not in moving it to the correct location afterwards. Reading the [documentation](https://docs.brew.sh/Formula-Cookbook#bininstall-foo), it seems like it ought to do just that.

**Reply by alebcay** [reply]
Can you confirm that the file being moved has the executable bit set? It's possible that `bin.install` is moving the file to the correct location, but then as part of the cleanup process, `brew` will remove anything from `bin` that is not executable.

**Reply by AngelaDMerkel** [reply]
It seems like the problem is that the binary depends on a variety of other files contained in the repo, mostly `.js` files which don't get moved into `/opt/homebrew/bin/` and therefore aren't available when I attempt to run. How can I ensure all the repo files are move into the `bin` where they belong

**Reply by branchv** [reply]
With `deno`, I think you'd have to manually move them yourself. Alternatively, I'd recommend just using `node` here:

```ruby
require "language/node"

class SingleFileCli < Formula  
  # ...
  depends_on "node"

  def install
    system "npm", "install", *Language::Node.std_npm_install_args(libexec)
    bin.install_symlink Dir["#{libexec}/bin/*"]
  end
```

**Reply by AngelaDMerkel** [reply]
Is it allowed to use another package manager like `npm` in `brew`?

**Reply by SMillerDev** [reply]
Yes, that is allowed
